artemis-server/src/main/java/org/apache/activemq/artemis/core/server/impl/QueueConsumersImpl.java

```diff
@@ -0,0 +1,648 @@
+      for (Level<T> level : levels) {
+         level.clear();
+      }
+      calcSize();
+   }
+
+
+
+   @Override
+   public boolean contains(Object o) {
+      return o instanceof PriorityAware && contains((PriorityAware) o);
+   }
+
+   public boolean contains(PriorityAware priorityAware) {
+      if (priorityAware == null) return false;
+      Level<T> level = getLevel(priorityAware.getPriority(), false);
+      return level != null && level.contains(priorityAware);
+   }
+
+   private void calcSize() {
+      Level<T>[] current = getArray();
+      int size = 0;
+      for (Level<T> level : current) {
+         size += level.size();
+      }
+      this.size = size;
+   }
+
+   private static class QueueConsumersIterator<T extends PriorityAware> implements ResetableIterator<T> {
+
+      private final QueueConsumersImpl<T> queueConsumers;
+      private final boolean resetable;
+      private Level<T>[] levels;
+      int level = -1;
+      private ResetableIterator<T> currentIterator;
+
+      private QueueConsumersIterator(QueueConsumersImpl<T> queueConsumers, boolean resetable) {
+         this.queueConsumers = queueConsumers;
+         this.levels = queueConsumers.getArray();
+         this.resetable = resetable;
+
+      }
+
+      @Override
+      public boolean hasNext() {
+         while (true) {
+            if (currentIterator != null) {
+               if (currentIterator.hasNext()) {
+                  return true;
+               }
+            }
+            int nextLevel = level + 1;
+            if (levels != null && nextLevel < levels.length) {
+               moveToLevel(nextLevel);
+            } else {
+               return false;
+            }
+         }
+      }
+
+      @Override
+      public T next() {
+         while (true) {
+            if (currentIterator != null) {
+               if (currentIterator.hasNext()) {
+                  return currentIterator.next();
+               }
+            }
+            int nextLevel = level + 1;
+            if (levels != null && nextLevel < levels.length) {
+               moveToLevel(nextLevel);
+            } else {
+               return null;
+            }
+         }
+      }
+
+      private void moveToLevel(int level) {
+         Level<T> level0 = levels[level];
+         if (resetable) {
+            currentIterator = level0.resetableIterator().reset();
+         } else {
+            currentIterator = level0.iterator();
+         }
+         this.level = level;
+      }
+
+      @Override
+      public ResetableIterator<T> reset() {
+         if (!resetable) {
+            throw new IllegalStateException("Iterator is not resetable");
+         }
+         levels = queueConsumers.getArray();
+         level = -1;
+         currentIterator = null;
+         return this;
+      }
+   }
+
+   /**
+    * This is represents a getPriority and is modeled on CopyOnArrayList.
+    *
+    * @param <E>
+    */
+   private static class Level<E> {
```

Comment:
I'm not a big fan of multiple layers of abstractions, but I admit that `Level` is sharing here a lot from `CopyOnWriteArratList`: we can't just reuse it somehow? It is due to the `setArray` call?

Reply:
No we cant, as we need access to getArray() without a copy, which is package access only in CopyOnWriteArrayList.